packages/adminrouter/extra/src/test-harness/tests/test_generic.py

```diff
@@ -0,0 +1,478 @@
+# Copyright (C) Mesosphere, Inc. See LICENSE file for details.
+
+import copy
+import logging
+import os
+
+import yaml
+
+import generic_test_code.common
+from generic_test_code.common import (
+    generic_correct_upstream_dest_test,
+    generic_correct_upstream_request_test,
+    generic_location_header_during_redirect_is_adjusted_test,
+    generic_no_slash_redirect_test,
+    generic_upstream_headers_verify_test,
+)
+
+log = logging.getLogger(__name__)
+
+
+def _merge_testconfig(a, b):
+    res = {'redirect_tests': [],
+           'location_header_rewrite_tests': [],
+           'endpoint_tests': [],
+           }
+
+    res['redirect_tests'].extend(
+        copy.deepcopy(a['redirect_tests']))
+    res['location_header_rewrite_tests'].extend(
+        copy.deepcopy(a['location_header_rewrite_tests']))
+    res['endpoint_tests'].extend(
+        copy.deepcopy(a['endpoint_tests']))
+
+    res['redirect_tests'].extend(
+        copy.deepcopy(b['redirect_tests']))
+    res['location_header_rewrite_tests'].extend(
+        copy.deepcopy(b['location_header_rewrite_tests']))
+    res['endpoint_tests'].extend(
+        copy.deepcopy(b['endpoint_tests']))
+
+    return res
+
+
+def _verify_tests_conf(tests_conf):
+    # TODO (prozlach): rewrite all of these to Json Schema validation, quick
+    # hack for now
+    _verify_endpoint_tests_conf(tests_conf['endpoint_tests'])
+    _verify_location_rewrite_tests(tests_conf['location_header_rewrite_tests'])
+    _verify_redirect_tests(tests_conf['redirect_tests'])
+
+
+def _verify_type_specification(types):
+    assert len(set(types)) in [1, 2]
+    for k in types:
+        assert k in ['master', 'agent']
+
+
+def _verify_redirect_tests(redirect_tests):
+    for t in redirect_tests:
+        assert len(t.keys()) == 2
```

Comment:
> JSON schema validator
Did not notice your second comment. This will not happen any time soon - I would love to have it, but it would take 1-2 days to do it right and there are lots of higher priority tasks in the queue.

I will refactor it the way you suggested and lets see if we can squeeze JSONSCHEMA validation somewhere. Thx!